### ISSUE TITLE
Instanceof left-hand-side check rejects generic index access types

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#19298]
Reported by @yangchristian in https://github.com/Microsoft/TypeScript/issues/2775#issuecomment-337463284

I repro @larsenwork's issue with the following:
```ts
type CreatedOrUpdatedAt =
    { createdAt: Date | string } |
    { updatedAt: Date | string };

const datePropComp = <T extends CreatedOrUpdatedAt, K extends keyof T>(a: T, b: T, field: K) => {
    // The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
    const dateA = a[field] instanceof Date ? a[field] : new Date(a[field]);
    const dateB = b[field] instanceof Date ? b[field] : new Date(b[field]);

    return dateA.getTime() - dateB.getTime();
};

datePropComp({ updatedAt: Date() }, { updatedAt: Date() }, 'updatedAt' )
```
[Playground link](https://www.typescriptlang.org/play/#src=type%20CreatedOrUpdatedAt%20%3D%0A%20%20%20%20%7B%20createdAt%3A%20Date%20%7C%20string%20%7D%20%7C%0A%20%20%20%20%7B%20updatedAt%3A%20Date%20%7C%20string%20%7D%3B%0A%0Aconst%20datePropComp%20%3D%20%3CT%20extends%20CreatedOrUpdatedAt%2C%20K%20extends%20keyof%20T%3E(a%3A%20T%2C%20b%3A%20T%2C%20field%3A%20K)%20%3D%3E%20%7B%0A%20%20%20%20%2F%2F%20The%20left-hand%20side%20of%20an%20'instanceof'%20expression%20must%20be%20of%20type%20'any'%2C%20an%20object%20type%20or%20a%20type%20parameter.%0A%20%20%20%20const%20dateA%20%3D%20a%5Bfield%5D%20instanceof%20Date%20%3F%20a%5Bfield%5D%20%3A%20new%20Date(a%5Bfield%5D)%3B%0A%20%20%20%20const%20dateB%20%3D%20b%5Bfield%5D%20instanceof%20Date%20%3F%20b%5Bfield%5D%20%3A%20new%20Date(b%5Bfield%5D)%3B%0A%0A%20%20%20%20return%20dateA.getTime()%20-%20dateB.getTime()%3B%0A%7D%3B%0A%0AdatePropComp(%7B%20updatedAt%3A%20Date()%20%7D%2C%20%7B%20updatedAt%3A%20Date()%20%7D%2C%20'updatedAt'%20))

**Comment by ryanlin1986** [comment]
This blocks us from using typescript 2.5, please fix it so we can upgrade.

**Comment by sandersn** [comment]
Duplicate of #17069 #17166 #15371

Fixed by #17912
